qiskit/pulse/library/symbolic_pulses.py

```diff
@@ -0,0 +1,763 @@
+                    else:
+                        constraints.append(sym.lambdify(params, constraint_expr))
+            ConstraintsDescriptor.global_constraints[clsname] = constraints
+            ConstraintsDescriptor.source_exprs[clsname] = source
+
+        return ConstraintsDescriptor.global_constraints[clsname]
+
+
+class SymbolicPulse(Pulse):
+    """The abstract superclass for parametric pulses.
+
+    Attributes:
+        PARAM_DEF (List[str]): A list of parameter names that constitutes a symbolic pulse.
+            Note that "t" and "limit" are reserved by the symbolic pulse superclass.
+            Subclass must define at least "duration" to define the length of instruction.
+        envelope (Callable): Lambdified symbolic equation to define the waveform envelope.
+            This attribute is populated when the subclass is instantiated first time.
+        constraints (List[Union[Callable, List]]): Lambdified symbolic equations to
+            provide validation for pulse parameters when assigned.
+            This attribute is populated when the subclass is instantiated first time.
+    """
+
+    __slots__ = ("param_values",)
+
+    PARAM_DEF = ["duration"]
+    envelope = EnvelopeDescriptor()
+    constraints = ConstraintsDescriptor()
+
+    @abstractmethod
+    def __init__(
+        self,
+        duration: Union[int, ParameterExpression],
+        parameters: Optional[Tuple[Union[ParameterExpression, complex], ...]] = None,
+        name: Optional[str] = None,
+        limit_amplitude: Optional[bool] = None,
+    ):
+        """Create a parametric pulse and validate the input parameters.
+
+        Args:
+            duration: Pulse length in terms of the sampling period `dt`.
+            parameters: Other parameters to form waveform.
+            name: Display name for this pulse envelope.
+            limit_amplitude: If ``True``, then limit the amplitude of the
+                waveform to 1. The default is ``True`` and the amplitude is constrained to 1.
+
+        Raises:
+            PulseError: When not all parameters are listed in the attribute :attr:`PARAM_DEF`.
+        """
+        super().__init__(duration=duration, name=name, limit_amplitude=limit_amplitude)
+
+        if parameters:
+            self.param_values = (duration,) + tuple(parameters)
+        else:
+            self.param_values = (duration,)
+
+        if len(self.param_values) != len(self.PARAM_DEF):
+            raise PulseError(
+                f"Number of parameters {len(self.param_values)} does not match "
+                f"with these defined parameters {self.PARAM_DEF}"
+            )
+
+        # Validate parameters when they are all assigned.
+        if not self.is_parameterized():
+            self.validate_parameters()
+
+            # This is awkward but likely necessary to populate data for QPY
+            # Perhaps we need another idea
+            if not self.envelope:
+                self.get_waveform()
```

Comment:
```suggestion
            # This is awkward but likely necessary to populate data for QPY
            # Perhaps we need another idea
            if not self.envelope:
                self.get_waveform()
```
Feel like this is bit awkward. This needs to be here because expression data is not populated until envelope function is called. In the actual use case, QPY dump may be called without visualizing schedule (which is one of situations get waveform is called), then, no definition of custom pulse is found and the program is saved without custom definition.

Reply:
I changed strategy in https://github.com/Qiskit/qiskit-terra/pull/7821/commits/43e0e0be1761f0a8de0c81dde02166f438f2655a

I added `definition` property that returns raw expression for envelope and constraints. Rather than storing the raw expression in descriptor, now I plan to ask QPY to call `definition` property to get them. This decouples the context of serialization from the descriptors and now they are storing only lambda functions. The original idea was let QPY check the descriptor class variable before investigating the pulse instructions so that we can add data block of custom pulse definitions before instruction data. However, the same thing can be done by temporary storing instruction data to IOByte buffer.